bindings/rust/generate/src/main.rs

```diff
@@ -0,0 +1,241 @@
+        })
+        .rust_target(bindgen::RustTarget::Stable_1_40)
+        // only export s2n-related stuff
+        .blocklist_type("iovec")
+        .blocklist_type("FILE")
+        .blocklist_type("_IO_.*")
+        .blocklist_type("__.*")
+        .blocklist_type("fpos_t")
+        // rust can't access thread-local variables
+        // https://github.com/rust-lang/rust/issues/29594
+        .blocklist_item("s2n_errno")
+        .rustified_enum("s2n_.*")
+        .raw_line(COPYRIGHT)
+        .raw_line(PRELUDE)
+        .ctypes_prefix("::libc")
+        .parse_callbacks(Box::new(S2nCallbacks::default()))
+        .clang_arg(format!("-I{}/api", s2n_dir.display()))
+        .clang_arg(format!("-I{}", s2n_dir.display()));
+    builder
+}
+
+fn gen_files(input: &Path, out: &Path) -> io::Result<()> {
+    use io::Write;
+
+    let mut files = std::fs::File::create(out)?;
+    let mut o = io::BufWriter::new(&mut files);
+
+    let pattern = format!("{}/**/*.c", input.display());
+
+    writeln!(o, "{}", COPYRIGHT)?;
+    writeln!(o, "[")?;
+    for file in glob::glob(&pattern).unwrap() {
+        let file = file.unwrap();
+        let file = file.strip_prefix(input).unwrap();
+        // don't include tests
+        if file.starts_with("tests") {
+            continue;
+        }
+        writeln!(o, "    {:?},", Path::new("lib").join(file).display())?;
+    }
+    writeln!(o, "]")?;
+    Ok(())
+}
+
+#[derive(Debug, Default)]
+struct S2nCallbacks;
+
+impl bindgen::callbacks::ParseCallbacks for S2nCallbacks {
+    fn enum_variant_name(
+        &self,
+        _enum_name: Option<&str>,
+        variant_name: &str,
+        _variant_value: bindgen::callbacks::EnumVariantValue,
+    ) -> Option<String> {
+        use heck::CamelCase;
+
+        if !variant_name.starts_with("S2N_") {
+            return None;
+        }
+
+        let variant_name = variant_name
+            .trim_start_matches("S2N_ERR_T_")
+            .trim_start_matches("S2N_EXTENSION_")
+            // keep the LEN_ so it's a valid identifier
+            .trim_start_matches("S2N_TLS_MAX_FRAG_")
+            .trim_start_matches("S2N_ALERT_")
+            .trim_start_matches("S2N_CT_SUPPORT_")
+            .trim_start_matches("S2N_STATUS_REQUEST_")
+            .trim_start_matches("S2N_CERT_AUTH_")
+            // match everything else
+            .trim_start_matches("S2N_");
+
+        Some(variant_name.to_camel_case())
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct FunctionCallbacks {
+    types: Arc<Mutex<BTreeSet<String>>>,
+    functions: Arc<Mutex<BTreeSet<String>>>,
+}
+
+impl FunctionCallbacks {
+    fn tests(&self, out: &Path) -> io::Result<()> {
+        use io::Write;
+        let functions = self.functions.lock().unwrap();
+        let mut types = self.types.lock().unwrap();
+
+        // bindgen doesn't have the ability to filter out type aliases
+        // so we'll need to have a list of them here for now
+        types.extend(
```

Comment:
will the CI job fail with a nice message if a new alias is added?

Reply:
Yes. It's not _the nicest_ message but it will block the build.

```
   Compiling s2n-tls-sys v0.1.0 (/home/cameron/Projects/awslabs/s2n/bindings/rust/s2n-tls-sys)
error[E0423]: expected value, found type alias `crate::s2n_verify_host_fn`
    --> src/tests.rs:1213:15
     |
1213 |     let ptr = crate::s2n_verify_host_fn as *const ();
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: can't use a type alias as a constructor
help: consider importing one of these items instead
     |
6    | use crate::tests::s2n_verify_host_fn;
     |

error: aborting due to previous error

For more information about this error, try `rustc --explain E0423`.
error: could not compile `s2n-tls-sys`
```